File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        
        vector<array<int,2>>adj[n];
        
        for(auto&f : flights){
            
            adj[f[0]].push_back({f[1],f[2]});
        }
        
        queue<array<int,2>>q;
        int dist[n];
        
        for(int i=0;i<n;i++){
            dist[i] = INT_MAX;
        }
        dist[src] = 0;
        
        q.push({src,0});
        int steps = 0;
        
        while(steps<=k and q.size()){
            
            int sz = q.size();
            
            while(sz --){
                
                array<int,2>arr= q.front();
                q.pop();
                
                //cout<<arr[0]<<" ";
                for(auto&v : adj[arr[0]]){
                    
                    int neigh = v[0], neighDist = v[1] + arr[1];
                    //cout<<neigh<<" , ";
                    if(dist[neigh]>neighDist){
                        dist[neigh] = neighDist;
                        q.push({neigh,neighDist});
                    }
                }
                //cout<<endl;
            }
            
            steps++;
        }
        
        
        return dist[dst] == INT_MAX? -1 : dist[dst];
    }
};

File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.java
class Solution {
    ArrayList<ArrayList<int[]>> adj;
    ArrayList<Integer> dist;
    Queue<int[]> q;

    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        adj = new ArrayList<>(n);
        dist = new ArrayList<>(n);
        
        // Initialize dist with Integer.MAX_VALUE
        for (int i = 0; i < n; i++) {
            dist.add(Integer.MAX_VALUE);
            adj.add(new ArrayList<>());
        }
        
        
        for (int[] flight : flights) {
            adj.get(flight[0]).add(new int[] { flight[1], flight[2] });
        }
        
        q = new LinkedList<>();
        q.offer(new int[]{src, 0});
        int steps =0;
        while (steps<=k && !q.isEmpty()) {
            int sz = q.size();
            while (sz-- > 0) {
                int[] curr = q.poll();
                for (int[] neigh : adj.get(curr[0])) {
                    int neighNode = neigh[0], neighDist = neigh[1] + curr[1];
                    if (dist.get(neighNode) > neighDist) {
                        dist.set(neighNode, neighDist);
                        q.offer(new int[]{neighNode, neighDist});
                    }
                }
            }
            steps++;
        }
        
        return dist.get(dst) == Integer.MAX_VALUE ? -1 : dist.get(dst);   
    }
}
